enhancements/machine-api/defaulting-validating.md

```diff
@@ -0,0 +1,137 @@
+- Externalise the providerSpec and expose a new CRD.
+- Flesh out the fields and details for each particular provider defaults and validation. That can be carefully discussed in each individual PR and updated back to this proposal in retrospective.
+
+## Proposal
+
+Today the machine API embeds the particular cloud provider API as a raw extension.
+
+```
+// ProviderSpec defines the configuration to use during node creation.
+type ProviderSpec struct {
+  Value *runtime.RawExtension `json:"value,omitempty"`
+}
+```
+
+This proposes to run both a `MutatingWebhookConfiguration` and a `ValidatingWebhookConfiguration` over the machine resource to let each provider define their API defaulting and validation.
+
+### User Stories [optional]
+
+#### Story 1
+
+As an operator I want the product to communicate meaningfully when my input is bad before it gets persisted so I can fix it and save time.
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Plumbing
+- The Machine API Operator (MAO) will expose both a `MutatingWebhookConfiguration` and a `ValidatingWebhookConfiguration` to the Cluster Version Operator (CVO).
+
+- The MAO will expose to the CVO a `Service` referenced in the webhook config that will resolve to the `Endpoint` where the webhook is running.
+
+- The webhook server can be managed by the same manager that manages the machineSet controller. This is a convenient place so it will let us manage single `MutatingWebhookConfiguration` and `ValidatingWebhookConfiguration` resources that can be extended with multiple `paths` for other machine API resources, e.g `machineSet`, `machineHealthCheck`, etc. All being reachable by the same `Service` resolving to the same `Endpoint`.
+
+- The serving certs and CA bundle for the webhook server will be managed by the https://github.com/openshift/service-ca-operator via `service.beta.openshift.io/serving-cert-secret-name:` and `"service.beta.openshift.io/inject-cabundle": "true"` annotations.
+
+#### Implementation
+
+Since most of the machine API controllers rely on controller-runtime we can leverage it to define the webhook server as well.
+We could implement `webhook.Defaulter` and `webhook.Validator` interfaces as in https://book.kubebuilder.io/cronjob-tutorial/webhook-implementation.html.
+
+However we need a more granular level of customisation which is not satisfied by those interfaces. Particularly we need to know the provider we are running on (to run the appropriate defaulting/validation) and the clusterID (to satisfy some defaults).
+To that end we can implement the [Handler interface](https://godoc.org/github.com/kubernetes-sigs/controller-runtime/pkg/webhook/admission#Handler) enabling us to infer the values above at the time of instantiating the webhook rather than on each request at runtime.
+
+### Risks and Mitigations
+
+As we introduce stronger opinions on defaults and validation we might break automated consumers that were relying on faulty behaviour.
```

Comment:
Is there something we can research via Telemetry or Insights to help us understand how many customers/clusters are affected with this "faulty" behavior?

Reply:
>Is there something we can research via Telemetry or Insights to help us understand how many customers/clusters are affected with this "faulty" behavior?

We don't have that granularity in telemetry atm. We are currently spiking to be able to answer some valuable questions out of it in next releases.